[RDY] Use buffered reading/writing for ShaDa files

ZyX-I opened the pull request:
Ref #4855

brcolow commented:
@Grimy Do you have time to run some benchmarks on this and compare it to the previous 9-byte per syscall? It would be great to compare.

brcolow commented on src/nvim/file.c:
```diff
@@ -0,0 +1,164 @@
+/// @param[in]  fname  File name to open.
+/// @param[in]  flags  Flags, @see FileOpenFlags.
+/// @param[in]  mode  Permissions for the newly created file (ignored if flags
+///                   does not have ).
```
What this a TODO? Flags does not have....what?

brcolow commented on src/nvim/file.c:
```diff
@@ -0,0 +1,164 @@
+/// @param[in]  fname  File name to open.
+/// @param[in]  flags  Flags, @see FileOpenFlags.
+/// @param[in]  mode  Permissions for the newly created file (ignored if flags
+///                   does not have ).
```
Same here.

oni-link commented on src/nvim/file.c:
```diff
@@ -0,0 +1,164 @@
+/// @param[in]  mode  Permissions for the newly created file (ignored if flags
+///                   does not have ).
+///
+/// @return [allocated] Opened file or
```
missing `NULL`

ZyX-I committed:
file,os/fs,shada: Separate opening, closing, writing and reading files

Moves low-level functions handling to os/fs.c. Adds file.c with a proxy
interface.

Target: while leaving syscalls handling is os.c (partially handled by libuv),
add buffering for reading and writing to file.c.

ZyX-I commented:
@brcolow This is WIP, no buffering, nothing to compare.

brcolow commented:
Oh darn...sorry about that.

ZyX-I commented:
Currently I just moved what was already there to os/fs.c and created a proxy interface in file.c where buffering will be added. Second one costs a `malloc()`/`free()` pair per ShaDa file plus both cost two levels of indirection (sd_\*->func() -> file_\*() -> os_\*() [-> uv_\*()] -> libc syscall wrapper; file_\* and os_\* did not exist, I also used uv_\* for closing and flushing file in os_\* while previously ShaDa code used functions from unistd.h directly; would’ve used for os_read and os_write as well, but `uv_read` is not something one wants to use if file is not supposed to support seeking and writing… why use libuv writing if not using libuv reading? Also documentation of libuv is written in a way so I may think that I need to handle EINTR myself, it also does not give any ideas about whether it will repeat writing if write was incomplete), so performance should actually be worse.

ZyX-I committed:
file: Add buffered reading and writing

Still no busted tests. Not tested without HAVE_PREADV.

ZyX-I commented:
I have the following test function:

```Lua
    local function os_read(fd, size)
      local buf = nil
      if size == nil then
        size = 0
      else
        buf = ffi.new('char[?]', size, ('\0'):rep(size))
      end
      local eof = ffi.new('bool[?]', 1, {true})
      local ret2 = fs.os_read(fd, eof, buf, size)
      local ret1 = eof[1]
      local ret3 = ''
      if buf ~= nil then
        ret3 = ffi.string(buf, size)
      end
      return ret1, ret2, ret3
    end
…
    describe('os_read', function()
      local file = 'test-unit-os-fs_spec-os_read.dat'

      local s = ''
      for i = 0, 255 do
        s = s .. (i == 0 and '\0' or ('%c'):format(i))
      end
      local fcontents = s:rep(16)

      before_each(function()
        local f = io.open(file, 'w')
        f:write(fcontents)
        f:close()
      end)

      after_each(function()
        os.remove(file)
      end)

      it('can read zero bytes from a file', function()
        local fd = os_open(file, ffi.C.kO_RDONLY, 0)
        ok(fd >= 0)
        eq({false, 0, ''}, {os_read(fd, nil)})
        eq({false, 0, ''}, {os_read(fd, 0)})
        eq(0, os_close(fd))
      end)

      it('can read from a file multiple times', function()
        local fd = os_open(file, ffi.C.kO_RDONLY, 0)
        ok(fd >= 0)
        eq({false, 2, '\000\001'}, {os_read(fd, 2)})
        eq({false, 2, '\002\003'}, {os_read(fd, 2)})
        eq(0, os_close(fd))
      end)

      it('can read the whole file at once and then report eof', function()
        local fd = os_open(file, ffi.C.kO_RDONLY, 0)
        ok(fd >= 0)
        eq({false, #fcontents, fcontents}, {os_read(fd, #fcontents)})
        eq({true, 0, ('\0'):rep(#fcontents)}, {os_read(fd, #fcontents)})
        eq(0, os_close(fd))
      end)
```

and it appears to set `eof` to false even though `printf` shows that `*ret_eof = true` was run correctly (and even if I have `*ret_eof = true` at the top of the function, though this fails `eq({false, 0, ''}, {os_read(fd, 0)})` - and only this, all other checks that rely on `false` are OK). Any ideas WTF is happening?

ZyX-I commented:
What is going on is “zero-based” indexing.

ZyX-I committed:
unittests: Add os_close, os_read and os_readv tests

justinmk commented on test/unit/os/fs_spec.lua:
```diff
@@ -432,55 +474,195 @@ describe('fs function', function()
         assert_file_exists(existing_file)
-        assert.is_true(0 <= (os_open(existing_file, ffi.C.kO_CREAT, 0)))
+        local fd = os_open(existing_file, ffi.C.kO_CREAT, 0)
+        assert.is_true(0 <= fd)
```
Test description says "returns non-negative", but this is testing for non-positive. I wonder if there was some drift.

ZyX-I replied:
@justinmk `0 <= fd` is non-negative, not non-positive. Just for some reason it uses confusing order (constant first, I would write `fd >= 0`), I did not change this.

justinmk commented:
@ZyX-I Do you plan to add `os_write` which wraps libuv? It may help with https://github.com/neovim/neovim/issues/4889.

ZyX-I commented:
@justinmk No, it can’t be used. According to the documentation it is equivalent to `pwritev` which needs file to be seekable and I have no reason to reject non-seekable files. And I do not see how reimplementing `os_write` on top of `uv_fs_write` and not regular `write` syscall is going to help with #4889: the only place where file name is used is `os_open` and this one is implemented on top of `uv_fs_open`.

ZyX-I commented:
By the way, why libuv does not have any sane `read` and `write` implementations? `p*v()` are inconvenient to use: you need to track file offset on your own, you cannot just put buffer in the arguments list and need temporary variable, these functions are absent in the standard (and I doubt that they are actual syscalls everywhere).

brcolow commented on src/nvim/os/fs.c:
```diff
@@ -325,23 +351,199 @@ static bool is_executable_in_path(const char_u *name, char_u **abspath)
+}
+
+#ifdef HAVE_READV
+/// Read from a file to a multiple buffers at once
```
The "a" is superfluous.

justinmk commented on src/nvim/file.c:
```diff
@@ -0,0 +1,293 @@
+///
+/// Buffered reading/writing to a file. Unlike fileio.c this is not dealing with
+/// Neovim stuctures for buffer, with autocommands, etc: just fopen/fread/fwrite
+/// replacement.
```
This should go in `os/`, no? That is where "[relatively] application-agnostic" operations live.

ZyX-I committed:
file: Use own constants, do not rely on fcntl.h

One of the reasons is that O_RDONLY is zero, which makes checking whether file
is opened read- or write-only harder. It is not guaranteed that on other system
O_WRONLY will not be zero (e.g. because file can only be opened in read-write
mode).

ZyX-I committed:
shada: Do not forget to close ShaDa reader

Previously there was file descriptor leak, not detected by sanitizers. Now it is 
file descriptor leak with a small memory leak which is detected by ASAN what 
fails one of the tests (actually, “ShaDa support code leaves .tmp.z in-place 
when there is error in original ShaDa and it has .tmp.a … .tmp.x”, but error is 
reported at the next test because leaks are not detected until Neovim exit and 
Neovim exit happens when clear()/reset() is called which happens in before_each 
only).

ZyX-I committed:
unittests: Add os_write test

New os/fs.c functions are now all tested.

ZyX-I commented:
New file_spec.lua tests SEGV somewhere in the garbage collector, typical backtrace is

```
#0  0x000000000042a4eb in lj_alloc_free (msp=0x40000010, ptr=0x40477a88) at lj_alloc.c:1311
        F = 0xffffffff00000000
        XP = 0xc1
        R = 0xfffffffb4000a1a8
        TP = 0x40477c00
        nsize = 18446744053308515832
        p = 0x40477a78
        fm = 0x40000010
        psize = 18446744053308516224
        next = 0x40477c00
#1  0x0000000000429af4 in lj_alloc_f (msp=0x40000010, ptr=0x40477a88, osize=384, nsize=0) at lj_alloc.c:1388
No locals.
#2  0x000000000043cf94 in lj_mem_free (g=0x400003b8, p=0x40477a88, osize=384) at ./lj_gc.h:120
No locals.
#3  resizetab (L=0x40000378, t=0x404776b8, asize=0, hbits=5) at lj_tab.c:273
        g = 0x400003b8
        i = 16
        oldnode = 0x40477a88
        oldasize = 0
        oldhmask = 15
#4  0x000000000043c7dc in rehashtab (L=0x40000378, t=0x404776b8, ek=0x7fffffffd390) at lj_tab.c:351
        bins = {0 <repeats 28 times>}
        total = 17
        asize = 0
        na = 0
        i = 28
#5  0x000000000043d68d in lj_tab_newkey (L=0x40000378, t=0x404776b8, key=0x7fffffffd390) at lj_tab.c:435
        nodebase = 0x40477a88
        collide = 0x40057400
        freenode = 0x40477a88
        n = 0x40477b30
#6  0x00000000004360fc in lj_BC_TSETS ()
No symbol table info available.
#7  0x000000000041437b in lua_pcall (L=0x40000378, nargs=7, nresults=-1, errfunc=2) at lj_api.c:1052
        g = 0x400003b8
        oldh = 0 '\000'
        ef = 24
        status = 0
#8  0x0000000000404aa1 in docall (L=0x40000378, narg=7, clear=0) at luajit.c:122
        status = 0
        base = 2
#9  0x0000000000404650 in handle_script (L=0x40000378, argv=0x7fffffffd678, n=9) at luajit.c:289
        status = 0
        fname = 0x7fffffffdccf "/home/zyx/.luarocks/lib/luarocks/rocks/busted/2.0.rc11-0/bin/busted"
        narg = 7
#10 0x0000000000403c86 in pmain (L=0x40000378) at luajit.c:538
        s = 0x6c8470 <smain>
        argv = 0x7fffffffd678
        script = 9
        flags = 12
#11 0x00000000004369e6 in lj_BC_FUNCC ()
No symbol table info available.
#12 0x0000000000414419 in lua_cpcall (L=0x40000378, func=0x403a60 <pmain>, ud=0x0) at lj_api.c:1074
        g = 0x400003b8
        oldh = 0 '\000'
        status = 1073742712
#13 0x0000000000403972 in main (argc=17, argv=0x7fffffffd678) at luajit.c:566
        status = 32767
        L = 0x40000378
```

Dunno what is the cause, but to fix this commenting file_read tests is needed.

ZyX-I committed:
unittests: Add tests for file.c

Also fixes some errors found.

ZyX-I commented:
Specifically I get problems if I comment all, but the first file_read tests.

ZyX-I commented:
Valgrind now proved to be not very useful: https://gist.github.com/d6335fb2c878ee3872490de9c1009b52.

tarruda commented:
> Valgrind now proved to be not very useful: https://gist.github.com/d6335fb2c878ee3872490de9c1009b52.

You can try to run the tests with plain lua:

```
LUA_DEBUGGER='valgrind --log-file=err.log' make functionaltest-lua
```

It is painful to debug segfaults in luajit, that's why I created `functionaltest-lua`

ZyX-I commented:
@tarruda Functional tests are fine (especially given that I have not written any new ones) and passing, so this is not an option. I would expect valgrind to show errors in *Neovim* code during unit tests, but this is not the case; also unit tests pass correctly if you are lucky.

Also as I was always saying early top-level makefile is doing many things I do not want build system to ever do, so I never use it. To run Valgrind I turned `~/.luarocks/bin/busted` into

```Shell
#!/bin/sh

trace="
    function _trace(event, line)
        local s = debug.getinfo(2).short_src
        io.stderr:write(s .. ':' .. line .. '\n')
    end
    debug.sethook(_trace, 'l')
"
trace=
preargs="gdbserver localhost:7777"
preargs="valgrind --leak-check=no --read-var-info=yes --track-origins=yes --num-callers=20 --trace-children=no --log-file=valgrind.log --tool=memcheck"
preargs=
exec $preargs 'luajit' -j off -e 'package.path="/home/zyx/.luarocks/share/lua/5.1/?.lua;/home/zyx/.luarocks/share/lua/5.1/?/init.lua;/usr/lib64/lua/luarocks/share/lua/5.1/?.lua;/usr/lib64/lua/luarocks/share/lua/5.1/?/init.lua;"..package.path; package.cpath="/home/zyx/.luarocks/lib/lua/5.1/?.so;/usr/lib64/lua/luarocks/lib/lua/5.1/?.so;"..package.cpath' -e 'local k,l,_=pcall(require,"luarocks.loader") _=k and l.add_context("busted","2.0.rc11-0")' -e "$trace" '/home/zyx/.luarocks/lib/luarocks/rocks/busted/2.0.rc11-0/bin/busted' "$@"
```

and edit which is more flexible (note: I have `-j off` (do not remember any occurrences of `jit` which BuildLua.cmake is stripping: from “luajit”?) and `$trace`) (and in any case I need to edit that to make busted use luajit in place of lua; not sure how to make it install with correct executable name from the very beginning, also AFAIR busted used to have absolute path there which is unacceptable because I have ~/.local-luajit -O0 -g build just for Neovim).

ZyX-I commented:
Basically if luajit was not crashing I could move the whole thing to `os/?.?` (not sure that this is the best place given that this is a wrapper around functions from `os/fs.c`, but still better then `src/nvim/file.c`), do some benchmarking\* and consider this ready. Though there is still question from #4547, if needed this may be done in a separate PR.

What is the best name for the `file.c`:

- `os/file.c`? Actually I used `file.c` because `fileio.c` was not available, this variant is also too generic because all I do here is providing buffered streams bound to file descriptors, not do some other file-related job (e.g. stat).
- `os/fileio.c`? I would go with this, but it may be confused with `fileio.c` from the top level.
- `os/io.c`? Too generic, I e.g. do not provide bufferring for sockets (though probably could: if I am not mistaking, the whole rbuffer.\* thing is for channel bufferring).
- Something else?

tarruda commented:
Oh sorry. I didn't read the whole thread and assumed it was the functional tests.

ZyX-I committed:
file: Move src/nvim/file.* to src/nvim/os/fileio.*

justinmk commented on src/nvim/os/fileio.h:
```diff
@@ -0,0 +1,72 @@
+#ifndef NVIM_FILE_H
+#define NVIM_FILE_H
```
This symbol should be changed too.

ZyX-I commented:
Some data:

```
(gdb) bt full
#0  0x000000000042a458 in lj_alloc_free (msp=0x40000010, ptr=0x400cb790) at lj_alloc.c:1311
        F = 0x42100000000
        B = 0x8cfe84182
        I = 0
        nsize = 0
        p = 0x400cb780
        fm = 0x40000010
        psize = 792
        next = 0x400cba98
#1  0x0000000000429af4 in lj_alloc_f (msp=0x40000010, ptr=0x400cb790, osize=784, nsize=0) at lj_alloc.c:1388
No locals.
#2  0x000000000047f8bc in lj_mem_free (g=0x400003b8, p=0x400cb790, osize=784) at ./lj_gc.h:120
No locals.
#3  lj_cdata_free (g=0x400003b8, cd=0x400cb798) at lj_cdata.c:75
No locals.
#4  0x0000000000439642 in gc_sweep (g=0x400003b8, p=0x400003e8, lim=2147482768) at lj_gc.c:405
        ow = 64
        o = 0x400cb798
#5  0x0000000000439475 in lj_gc_freeall (g=0x400003b8) at lj_gc.c:560
        i = 0
        strmask = 1074301736
#6  0x000000000040d11c in close_state (L=0x40000378) at lj_state.c:159
        g = 0x400003b8
#7  0x000000000040d4a5 in lua_close (L=0x40000378) at lj_state.c:258
        g = 0x400003b8
        i = 1
#8  0x0000000000403992 in main (argc=18, argv=0x7fffffffd638) at luajit.c:568
        status = 0
        L = 0x40000378
(gdb) frame 3
#3  lj_cdata_free (g=0x400003b8, cd=0x400cb798) at lj_cdata.c:75
75          lj_mem_free(g, memcdatav(cd), sizecdatav(cd));
(gdb) p *cd
$22 = {nextgc = {gcptr32 = 1074685256}, marked = 129 '\201', gct = 10 '\n', ctypeid = 116}
(gdb) p (*((CTState *)(uintptr_t)(g->ctype_state.ptr32))).tab[116]
$23 = {info = 806354949, size = 4294967295, sib = 0, next = 0, name = {gcptr32 = 0}}
(gdb) p (806354949 >> 28)
$24 = 3
```

If I correctly understood what that means, problem occurs with some array (another + for the version that this has something to do with `char[?]` array: it is the only thing `file_read` is allocating via `ffi.new`). But size is rather strange, both what I see in `size` in ctype tab (though this may be yet another “store as much data in one integer as you can” trick) and what is passed to `lj_mem_free`. Maybe this is just “768 + some more space for luajit needs”.

ZyX-I commented:
Now interesting: it *stops crashing* if I use

        buf = ffi.new('char[?]', size, ('\0'):rep(size - 1))

or

        buf = ffi.new('char[?]', size + 1, ('\0'):rep(size))

in place of

        buf = ffi.new('char[?]', size, ('\0'):rep(size))

This is rather strange because `os_read` and `os_readv` are also tested with size matching string length and that does not crash.

ZyX-I committed:
unittests: Fix bug somewhere that makes file_read tests SEGV

ZyX-I committed:
*: Satisfy linter (newest type casts rule)

ZyX-I committed:
unittests: Fix kFileNoSymlink test on FreeBSD

Actual value on FreeBSD is -31, UV_EMLINK was obtained from
/usr/include/asm-generic/errno-base.h (there EMLINK is defined as 31 there).
This may actually be something else, but I do not think so as “Too many links”
description also fits in. [Man page][1] agrees with me, search for `[EMLINK]`
([linux man page][2] also specifies ELOOP explicitly in a similar section).

[1]: https://www.freebsd.org/cgi/man.cgi?query=open&sektion=2
[2]: http://man7.org/linux/man-pages/man3/open.3p.html

ZyX-I commented:
Some benchmarks:

```
% ./scripts/shadacat.py ~/.local/share/nvim/shada/main.shada
   0        Header 2016-06-23T20:32:32    85 {'max_kbyte': 10, 'generator': 'nvim', 'encoding': 'utf-8', 'pid': 4237, 'version': 'NVIM v0.1.5-495-gbafa013'}
  92          Jump 2016-06-08T21:50:15    84 {'f': '/home/zyx/a.a/Proj/c/neovim/build/root/share/nvim/runtime/doc/nvim_provider.txt'}
 183          Jump 2016-06-08T21:51:35    87 {'f': '/home/zyx/a.a/Proj/c/neovim/build/root/share/nvim/runtime/doc/nvim_provider.txt', 'l': 2}
 277          Jump 2016-06-08T21:51:41    87 {'c': 3, 'f': '/home/zyx/a.a/Proj/c/neovim/build/root/share/nvim/runtime/plugin/rplugin.vim', 'l': 7}
 371          Jump 2016-06-08T21:51:54    84 {'f': '/home/zyx/a.a/Proj/c/neovim/build/root/share/nvim/runtime/plugin/rplugin.vim', 'l': 6}
 462     SubString 1970-01-01T03:00:00     3 ['']
 468     LocalMark 2016-06-08T22:00:33    94 {'c': 4, 'f': '/home/zyx/a.a/Proj/c/neovim/build/root/share/nvim/runtime/autoload/remote/host.vim', 'l': 161}
 569     LocalMark 2016-06-08T21:51:54    84 {'f': '/home/zyx/a.a/Proj/c/neovim/build/root/share/nvim/runtime/plugin/rplugin.vim', 'l': 6}
 660     LocalMark 2016-06-08T21:51:40    90 {'c': 4, 'l': 7, 'f': '/home/zyx/a.a/Proj/c/neovim/build/root/share/nvim/runtime/plugin/rplugin.vim', 'n': 94 ('^')}
 757     LocalMark 2016-06-08T21:51:41    87 {'n': 46 ('.'), 'f': '/home/zyx/a.a/Proj/c/neovim/build/root/share/nvim/runtime/plugin/rplugin.vim', 'l': 7}
 851        Change 2016-06-08T21:51:41    84 {'f': '/home/zyx/a.a/Proj/c/neovim/build/root/share/nvim/runtime/plugin/rplugin.vim', 'l': 7}
 942     LocalMark 2016-06-08T21:51:35    87 {'f': '/home/zyx/a.a/Proj/c/neovim/build/root/share/nvim/runtime/doc/nvim_provider.txt', 'l': 2}
1036  HistoryEntry 2016-06-08T21:50:15    51 [0, 'e root/share/nvim/runtime/doc/nvim_provider.txt']
1094  HistoryEntry 2016-06-08T21:51:20    26 [0, 'h :UpdateRemotePlugins']
1127  HistoryEntry 2016-06-08T21:51:27    36 [0, 'verb command UpdateRemotePlugins']
1170  HistoryEntry 2016-06-08T21:51:35    48 [0, 'e root/share/nvim/runtime/plugin/rplugin.vim']
1225  HistoryEntry 2016-06-08T21:51:54    54 [0, 'e root/share/nvim/runtime/autoload/remote/host.vim']
1286  HistoryEntry 2016-06-22T15:55:24    31 [0, "echo capture('silent! foo')"]
1324  HistoryEntry 2016-06-22T15:57:00    71 [0, "try | capture('silent! foo') | catch | echomsg v:exception | endtry"]
1402  HistoryEntry 2016-06-22T15:57:09    76 [0, "try | echo capture('silent! foo') | catch | echomsg v:exception | endtry"]
1485  HistoryEntry 2016-06-22T15:57:16    68 [0, "try | echo capture('foo') | catch | echomsg v:exception | endtry"]
1560  HistoryEntry 2016-06-23T20:32:31     6 [0, 'qa']
```

```
% cat ~/.local/share/nvim/shada/main.shada{,}{,}{,}{,}{,}{,}{,}{,}{,}{,}{,}{,} > /tmp/test.shada
% la /tmp/test.shada
-rw-r--r-- 1 zyx 6.2M Jun 24 00:16 /tmp/test.shada
```

```
% mkdir build
% cd build
% CC=gcc cmake .. -DCMAKE_BUILD_TYPE=Release
% make
```

1. My branch (6580dfeddd9693b778f59ca5c333d05200d67687):

```
% strace -o /dev/stdout -e read,readv bin/nvim -u NONE -i /tmp/test.shada --cmd 'let rt=reltime()' --cmd 'rshada!' --cmd 'echo reltimestr(reltime(rt))' --cmd 'set shada=' --cmd cq 2>/dev/null | wc -l
6321
% bin/nvim -u NONE -i /tmp/test.shada --cmd 'let rt=reltime()' --cmd 'rshada!' --cmd 'echo reltimestr(reltime(rt))' --cmd 'set shada=' --cmd cq
  0.066031
```

2. master (65af001f2bcc35f19d64b4d2c1dbcd46d87432e8):

```
% strace -o /dev/stdout -e read,readv bin/nvim -u NONE -i /tmp/test.shada --cmd 'let rt=reltime()' --cmd 'rshada!' --cmd 'echo reltimestr(reltime(rt))' --cmd 'set shada=' --cmd cq 2>/dev/null | wc -l
446493
% bin/nvim -u NONE -i /tmp/test.shada --cmd 'let rt=reltime()' --cmd 'rshada!' --cmd 'echo reltimestr(reltime(rt))' --cmd 'set shada=' --cmd cq
  0.202231
```

Note: benchmark is constructed in a way (tmpfs + multiple calls so file is cached), so that syscalls themeselves are a problem, not something else like disk IO. In any case I would expect `main.shada` to live in a cache. As you see, under this circumstances number of syscalls was reduced very significantly (446493 / 6321 ≈ 70.6), time was also reduced, but rate is not as good here: 0.202231 / 0.066031 ≈ 3.06.

ZyX-I commented:
Note though that strace failed to catch any occurrences of `readv`: to get it used one needs larger ShaDa entries. I will now try to check whether using `readv` under different conditions will make code faster.

ZyX-I commented:
I have not noticed any performance improvement in 96a57e1bc66f294684e9ac2ee8f2b74243982123, but number of syscalls was reduced to 6173. Also I fixed errors found in `#if HAVE_READV` block.

ZyX-I commented:
If tests pass this may be considered RDY, assuming answer on question from #4547 is negative or it is OK to move implementation of positive answer to next PR. In any case to make positive answer useful for resolving #4547 `undo.c` will need to be refactored to use `os/fileio.c` in place of `stdio.h`.

ZyX-I committed:
os/fileio: Use readv often

ZyX-I committed:
unittests: Fix testlint errors

justinmk commented:
IMO very in favor of your suggestion in #4547

justinmk commented:
This PR is a fantastic improvement. Results from a particularly slow 32-bit environment with a 124KB main.shada:

#### Before:
```
time nvim  -u NORC +'call feedkeys("ZQ")'
real    0m0.966s
user    0m0.012s
sys     0m0.884s
```

#### After:

```
time nvim  -u NORC +'call feedkeys("ZQ")'
real    0m0.291s
user    0m0.012s
sys     0m0.204s
```

ZyX-I commented:
I think, this is RDY: no need to delay finished work for #4547 which may be fixed separately.

tarruda commented:
> buf = ffi.new('char[?]', size, ('\0'):rep(size))

I think this is the same issue I had last year while writing rbuffer tests: http://www.freelists.org/post/luajit/Segfault-with-32bit-luajit-ffi-and-some-C-code,1

When you pass `char[?]` it is interpreted as VLA which doesn't have a fixed size, so it will always take the size from the lua string plus NUL. `size` parameter is ignored.

ZyX-I commented:
@tarruda Given how I fixed this, it definitely is not ignored. It looks like some code uses one size (given by me as a second argument) and some code uses another (derived from string length), if it was ignored I would not be seeing any crashes because length of the third argument + 1 is absolutely fine size for my purposes (so I do not overflow) and luajit code ignoring the second argument will not overflow as well (as it only knows one fine size).

oni-link commented on src/nvim/os/fileio.c:
```diff
@@ -0,0 +1,319 @@
+/// Write to a file
+///
+/// @param[in]  fd  File descriptor to write to.
+/// @param[in]  buf  Data to write. May be NULL if size is zero.
```
Source buffer for `rbuffer_write()` is expected to be not `NULL`. A size of zero also cannot be handled because `rbuffer_produced()` is called with that value.
